File: backend/app/properties/schemas.py
"""
Properties Module — Schemas
"""

from pydantic import BaseModel, Field

from app.properties.models import ListingPlatform, PropertyStatus, PropertyType
from app.shared.schemas import BaseResponse


class PropertyCreateRequest(BaseModel):
    name: str
    type: PropertyType = PropertyType.OTHER
    address: str | None = None
    city: str | None = None
    state: str | None = None
    country: str | None = None
    postal_code: str | None = None
    bedrooms: int = 1
    bathrooms: float = 1.0
    max_guests: int = 2
    square_meters: float | None = None
    acquisition_cost: float | None = None
    monthly_mortgage: float | None = None
    target_occupancy: float | None = None
    target_annual_revenue: float | None = None
    notes: str | None = None


class PropertyUpdateRequest(BaseModel):
    name: str | None = None
    type: PropertyType | None = None
    status: PropertyStatus | None = None
    address: str | None = None
    city: str | None = None
    state: str | None = None
    country: str | None = None
    postal_code: str | None = None
    bedrooms: int | None = None
    bathrooms: float | None = None
    max_guests: int | None = None
    square_meters: float | None = None
    acquisition_cost: float | None = None
    monthly_mortgage: float | None = None
    target_occupancy: float | None = None
    target_annual_revenue: float | None = None
    notes: str | None = None


class ListingResponse(BaseResponse):
    platform: ListingPlatform
    external_id: str | None = None
    listing_url: str | None = None
    is_active: bool
    base_price: float | None = None
    currency: str


class PropertyResponse(BaseResponse):
    name: str
    type: PropertyType
    status: PropertyStatus
    address: str | None = None
    city: str | None = None
    state: str | None = None
    country: str | None = None
    postal_code: str | None = None
    bedrooms: int
    bathrooms: float
    max_guests: int
    square_meters: float | None = None
    acquisition_cost: float | None = None
    monthly_mortgage: float | None = None
    target_occupancy: float | None = None
    target_annual_revenue: float | None = None
    notes: str | None = None


class PropertyDetailResponse(PropertyResponse):
    """Full property response with listings (eagerly loaded)."""
    listings: list[ListingResponse] = Field(default_factory=list)


class PropertySummaryResponse(BaseResponse):
    """Lightweight property representation for lists."""
    name: str
    type: PropertyType
    status: PropertyStatus
    city: str | None = None
    country: str | None = None
    bedrooms: int
    max_guests: int
    listing_count: int = 0


class ListingCreateRequest(BaseModel):
    platform: ListingPlatform
    external_id: str | None = None
    listing_url: str | None = None
    base_price: float | None = None
    currency: str = "USD"
    minimum_nights: int = 1
    maximum_nights: int | None = None

File: app/backend/src/properties/schemas.rs
//! Property request schemas, mirroring `backend/app/properties/schemas.py`.

use serde::Deserialize;

fn default_type() -> String {
    "other".into()
}
fn default_status() -> String {
    "active".into()
}
fn default_bedrooms() -> i64 {
    1
}
fn default_bathrooms() -> f64 {
    1.0
}
fn default_max_guests() -> i64 {
    2
}

#[derive(Debug, Deserialize)]
pub struct PropertyCreateRequest {
    pub name: String,
    #[serde(default = "default_type", rename = "type")]
    pub r#type: String,
    #[serde(default = "default_status")]
    pub status: String,
    pub address: Option<String>,
    pub city: Option<String>,
    pub state: Option<String>,
    pub country: Option<String>,
    pub postal_code: Option<String>,
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
    #[serde(default = "default_bedrooms")]
    pub bedrooms: i64,
    #[serde(default = "default_bathrooms")]
    pub bathrooms: f64,
    #[serde(default = "default_max_guests")]
    pub max_guests: i64,
    pub square_meters: Option<f64>,
    pub acquisition_cost: Option<f64>,
    pub monthly_mortgage: Option<f64>,
    pub target_occupancy: Option<f64>,
    pub target_annual_revenue: Option<f64>,
    pub notes: Option<String>,
}

/// PATCH semantics: only fields present in the JSON are updated.
#[derive(Debug, Default, Deserialize)]
pub struct PropertyUpdateRequest {
    pub name: Option<String>,
    #[serde(rename = "type")]
    pub r#type: Option<String>,
    pub status: Option<String>,
    pub address: Option<String>,
    pub city: Option<String>,
    pub state: Option<String>,
    pub country: Option<String>,
    pub postal_code: Option<String>,
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
    pub bedrooms: Option<i64>,
    pub bathrooms: Option<f64>,
    pub max_guests: Option<i64>,
    pub square_meters: Option<f64>,
    pub acquisition_cost: Option<f64>,
    pub monthly_mortgage: Option<f64>,
    pub target_occupancy: Option<f64>,
    pub target_annual_revenue: Option<f64>,
    pub notes: Option<String>,
}

#[derive(Debug, Default, Deserialize)]
pub struct ListParams {
    pub skip: Option<i64>,
    pub limit: Option<i64>,
}
